server/src-lib/Hasura/App.hs

```diff
@@ -270,15 +270,15 @@ newShutdownLatch = fmap ShutdownLatch C.newEmptyMVar
 
 -- | Block the current thread, waiting on the latch.
 waitForShutdown :: ShutdownLatch -> IO ()
-waitForShutdown = C.takeMVar . unShutdownLatch
+waitForShutdown = C.readMVar . unShutdownLatch
 
 -- | Initiate a graceful shutdown of the server associated with the provided
 -- latch.
 shutdownGracefully :: InitCtx -> IO ()
 shutdownGracefully = shutdownGracefully' . _icShutdownLatch
 
 shutdownGracefully' :: ShutdownLatch -> IO ()
-shutdownGracefully' = flip C.putMVar () . unShutdownLatch
+shutdownGracefully' = void . flip C.tryPutMVar () . unShutdownLatch
```

Comment:
I think it would be good to return the result instead of discarding it. Trying to shut down twice should result in an error in some cases IMO.

Reply:
ok

Reply:
I don't call void here any more.